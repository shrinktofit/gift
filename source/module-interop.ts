
import ts from 'typescript';

export class ModuleInteropRecord {
    constructor(name: string) {
        this._myName = name;
    }

    get record(): Readonly<ModuleInteropRecord['_record']> {
        return this._record;
    }

    get selfExports() {
        return this._selfExports;
    }

    public addNamedImport(from: string, importName: string, exportName: string) {
        const interop = this._getInterop(from);
        const isImportExportNameSame = importName === exportName;
        for (const { propertyName, name } of interop.imports) {
            if (isImportExportNameSame) {
                if (name.text === importName) {
                    return;
                }
            } else {
                if (propertyName && propertyName.text === importName &&
                    name.text === exportName) {
                    return;
                }
            }
        }
        interop.imports.push(isImportExportNameSame ?
            ts.createImportSpecifier(undefined, ts.createIdentifier(importName)):
            ts.createImportSpecifier(ts.createIdentifier(exportName), ts.createIdentifier(importName)),
        );
    }

    public addNamedExportFrom(from: string, importName: string, exportName: string) {
        const interop = this._getInterop(from);
        const isImportExportNameSame = importName === exportName;
        if (!interop.exports.some((specifier) => this._isEqualExportSpecifier(specifier, importName, exportName))) {
            interop.exports.push(isImportExportNameSame ?
                ts.createExportSpecifier(undefined, ts.createIdentifier(importName)):
                ts.createExportSpecifier(ts.createIdentifier(exportName), ts.createIdentifier(importName)),
            );
        }
    }
    
    public addNamedExport(localName: string, exportName: string) {
        if (!this._selfExports.some((specifier) => this._isEqualExportSpecifier(specifier, localName, exportName))) {
            this._selfExports.push((localName === exportName) ?
                ts.createExportSpecifier(undefined, ts.createIdentifier(localName)):
                ts.createExportSpecifier(ts.createIdentifier(exportName), ts.createIdentifier(localName)),
            );
        }
    }

    private _myName: string;
    private _selfExports: ts.ExportSpecifier[] = [];
    private _record: Map<string, {
        specifier: string;
        imports: ts.ImportSpecifier[];
        exports: ts.ExportSpecifier[];
    }> = new Map();

    private _getInterop(from: string) {
        let interop = this._record.get(from);
        if (!interop) {
            const specifier = this._optimizeModuleSpecifierTo(from);
            interop = {
                specifier,
                imports: [],
                exports: [],
            };
            this._record.set(from, interop);
        }
        return interop;
    }

    private _optimizeModuleSpecifierTo(to: string): string {
        return to;
    }

    private _isEqualExportSpecifier(
        specifier: ts.ExportSpecifier, localOrImportName: string, exportName: string) {
        if (specifier.name.text !== localOrImportName) {
            return false;
        }
        if (localOrImportName === exportName) {
            return !specifier.propertyName || specifier.propertyName.text === exportName;
        } else {
            return specifier.propertyName && specifier.propertyName.text === exportName;
        }
    }
}
