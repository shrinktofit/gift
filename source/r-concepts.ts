import ts from 'typescript';

const noParent: NamespaceTraits = null as unknown as NamespaceTraits;

export class Entity {
    /**
     * `null` indicates this entity is created by us.
     */
    public symbol: ts.Symbol | null;

    public moduleTraits?: ModuleTraits;

    constructor(parent: NamespaceTraits, name: string, symbol: ts.Symbol | null) {
        this.symbol = symbol;
        this._name = name;
        if (parent === noParent) {
            this._fullPath = [ ];
        } else {
            if (parent.entity.isModule) {
                this._ownerModule = parent.entity.namespaceTraits!;
            } else {
                this._ownerModule = parent.entity.ownerModule;
            }
            this._fullPath = parent.entity.fullPath.slice();
            if (symbol) {
                parent.children.push(this);
            }
        }
        this._fullPath.push(this);
    }

    get isModule() {
        return !!this.moduleTraits;
    }

    get name() {
        return this._name;
    }

    get parent() {
        return this._fullPath[this._fullPath.length - 2].namespaceTraits!;
    }

    get fullPath() {
        return this._fullPath;
    }

    get ownerModule(): NamespaceTraits {
        return this._ownerModule;
    }

    get namespaceTraits() {
        return this._namespaceTraits;
    }

    public addNamespaceTraits() {
        this._namespaceTraits = {
            entity: this,
            children: [],
            aliasExports: [],
        };
        return this._namespaceTraits;
    }

    private _name: string;
    private declare _ownerModule: NamespaceTraits;
    private _namespaceTraits?: NamespaceTraits;
    private _fullPath: Entity[] = [];
}

export interface BaseTraits {
    entity: Entity;
}

export interface ModuleTraits extends BaseTraits {
    imports: Record<string, ImportDetail>;
}

export interface ImportDetail {
    namedImports: Record<string, string>;
}

export interface NamespaceTraits extends BaseTraits {
    children: Entity[];
    aliasExports: Array<{
        module: NamespaceTraits;
        importName: string;
        exportName: string;
    }>;
    neNamespace?: {
        name: string;
        statements: ts.Statement[];
    };
}

export function createModule(name: string, symbol: ts.Symbol) {
    const entity = new Entity(noParent, name, symbol);
    entity.addNamespaceTraits();
    entity.moduleTraits = {
        entity,
        imports: {},
    };
    return entity;
}
