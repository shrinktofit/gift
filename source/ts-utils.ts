
import ts from 'typescript';

const nodeFactory = ts.factory;

export function splitLeftmost(entityName: ts.EntityName) {
    let rights: ts.Identifier[] = [];
    let leftmost: ts.EntityName = entityName;
    while (true) {
        if (ts.isIdentifier(leftmost)) {
            break;
        }
        rights.unshift(leftmost.right);
        leftmost = leftmost.left;
    }
    return {
        leftmost,
        rights,
    };
}

export function createEntityName(identifiers: string[], leftmost: ts.EntityName | null = null): ts.EntityName {
    let result = leftmost;
    for (const id of identifiers) {
        const newID = nodeFactory.createIdentifier(id);
        if (!result) {
            result = newID;
        } else {
            result = nodeFactory.createQualifiedName(result, newID);
        }
    }
    return result!;
}

export function createAccessLink(identifiers: string[], leftmost: ts.Expression | null = null): ts.Expression {
    let result = leftmost;
    for (const id of identifiers) {
        const newID = nodeFactory.createIdentifier(id);
        if (!result) {
            result = newID;
        } else {
            result = ts.createPropertyAccess(result, newID);
        }
    }
    return result!;
}

export function printSymbol(symbol: ts.Symbol) {
    const declaration = symbol.valueDeclaration || (
        (symbol.declarations !== undefined && symbol.declarations.length !== 0) ? symbol.declarations[0] : null);
    console.log(
        `[[${symbol.name}]], \n` +
        `  ${declaration ? stringifyNode(declaration) : '!!NO-DECLARATION!!'}, \n` +
        `  ${stringifySymbolFlags(symbol.flags)}`);
}

export function stringifySymbolFlags(flags: ts.SymbolFlags) {
    const satisfies: string[] = [];
    for (const key of Object.keys(ts.SymbolFlags)) {
        if (typeof key === 'string') {
            const value = (ts.SymbolFlags as any)[key];
            if (flags & value) {
                satisfies.push(key);
            }
        }
    }
    return satisfies.join(',');
}

export function stringifyNode(node: ts.Node) {
    return `Syntax Kind: ${ts.SyntaxKind[node.kind]}`;
}