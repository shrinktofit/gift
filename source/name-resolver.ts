import * as rConcepts from './r-concepts';

export class NameResolver {
    public resolve(entity: rConcepts.Entity) {
        return resolveRelativePath(this.current(), entity);
    }

    public enter(ns: rConcepts.NamespaceTraits) {
        console.log(`++ ${printNamespace(ns)}`);
        this._contextStack.push(ns);
    }

    public leave() {
        if (this._contextStack.length === 0) {
            throw new Error(`Bad NameResolver.leave() call`);
        } else {
            console.log(`-- ${printNamespace(this.current())}`);
            this._contextStack.pop();
        }
    }

    public current() {
        if (this._contextStack.length === 0) {
            // Shall not happen
            debugger;
        }
        const context = this._contextStack[this._contextStack.length - 1];
        return context;
    }
    
    private _contextStack: rConcepts.NamespaceTraits[] = [];
}

export namespace NameResolver {
    export interface ResolveResult {
        module?: string;
        namespaces?: string[];
        name: string;
    }
}

function printNamespace(trait: rConcepts.NamespaceTraits) {
    const fullPath = trait.entity.fullPath;
    return `"${fullPath[0].name}"${fullPath.length === 0 ? '' : `.${fullPath.slice(1).map(e => e.name).join('.')}`}`;
}

export function resolveRelativePath(from: rConcepts.NamespaceTraits, to: rConcepts.Entity) {
    const fromFullPath = from.entity.fullPath;
    const toFullPath = to.fullPath;

    let iUnmatchedNamespace = 0;
    for (; iUnmatchedNamespace < fromFullPath.length &&
        iUnmatchedNamespace < toFullPath.length;
        ++iUnmatchedNamespace) {
        if (fromFullPath[iUnmatchedNamespace] !== toFullPath[iUnmatchedNamespace]) {
            break;
        }
    }

    // assert(toFullPath.length >= 2);
    const result: NameResolver.ResolveResult = {
        name: toFullPath[toFullPath.length - 1].name,
    };

    // We should consider an edge case:
    // an inner namespace 'A' reference outer namespace 'B', for example,
    // the shortest path is "a.b....".
    // However the 'A' may also has a (nested) namespace named "a.b....".
    // So the conflict occurs.
    // In such case, we have to prefer the longest path to avoid conflict.
    let iNoConflict = iUnmatchedNamespace;
    if (iNoConflict < toFullPath.length) {
        while (iNoConflict > 0) {
            const entity = toFullPath[iNoConflict];
            if (!mayUnambiguousReferenceTo(entity, from, entity.name)) {
                --iNoConflict;
            } else {
                break;
            }
        }
    }

    if (iNoConflict === 0) {
        // Module mismatch
        result.module = toFullPath[0].name;
        ++iNoConflict;
    }

    const sliceSize = toFullPath.length - iNoConflict - 1;
    if (sliceSize > 0) {
        result.namespaces = toFullPath.slice(iNoConflict, iNoConflict + sliceSize).map(n => n.name);
    }

    return result;
}

function mayUnambiguousReferenceTo(ref: rConcepts.Entity, from: rConcepts.NamespaceTraits, name: string): boolean {
    let ns: rConcepts.NamespaceTraits = from;
    while (true) {
        for (const bro of ns.children) {
            if (bro === ref) {
                return true;
            } else if (bro.name === name) {
                return false;
            }
        }
        if (ns.entity.isModule) {
            break;
        } else {
            ns = ns.entity.parent;
        }
    }
    return true;
}