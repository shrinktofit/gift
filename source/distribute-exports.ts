import ts from 'typescript';
import { hasJsDocTag } from './ts-utils';

export function distributeExports(
    moduleSymbols: ts.Symbol[],
    typeChecker: ts.TypeChecker,
    priorityList: string[] = [],
    privateJsDocTag?: string,
) {
    const parsedPriorityList = priorityList.map((id) => `"${id.replace(/\\/g, '/').replace(/\.(js|ts|d\.ts)$/, '')}"`);

    const exportMap: Map<ts.Symbol, SymbolInfo> = new Map();

    const moduleMetaList = moduleSymbols.map((moduleSymbol) => {
        const moduleMeta: distributeExports.ModuleMeta = {
            symbol: moduleSymbol,
            mainExports: [],
            aliasExports: [],
            [prioritySymbol]: getExportPriority(moduleSymbol, parsedPriorityList),
        };
        iterateModuleExports(moduleSymbol, moduleMeta);
        return moduleMeta;
    });

    exportMap.forEach((symbolInfo, originalSymbol) => {
        const { exportPorts } = symbolInfo;
        const iMainExportPort = findBestExportMeta(originalSymbol, exportPorts);
        const mainExportPort = exportPorts[iMainExportPort];
        const mainExportModule = mainExportPort.module;
        const mainExportIndex = mainExportModule.mainExports.length;
        mainExportModule.mainExports.push({
            originalSymbol,
            exportSymbol: mainExportPort.through,
            children: symbolInfo.children,
        });
        exportPorts.forEach((aliasingExportPort, iAliasingExportPort) => {
            if (iAliasingExportPort === iMainExportPort) {
                return;
            }
            aliasingExportPort.module.aliasExports.push({
                module: mainExportModule,
                mainExportIndex: mainExportIndex,
                exportSymbol: aliasingExportPort.through,
            });
        });
    });

    return moduleMetaList;

    interface ExportPort {
        through: ts.Symbol;
        module: distributeExports.InternalModuleMeta;
    }

    interface SymbolInfo {
        exportPorts: ExportPort[];

        /**
         * Exists if this is a module symbol.
         */
        children?: distributeExports.InternalModuleMeta[];
    }

    function iterateModuleExports(moduleSymbol: ts.Symbol, moduleMeta: distributeExports.InternalModuleMeta) {
        const exportSymbols = typeChecker.getExportsOfModule(moduleSymbol);
        for (const exportedSymbol of exportSymbols) {
            let originalSymbol = exportedSymbol;
            if (exportedSymbol.flags & ts.SymbolFlags.Alias) {
                originalSymbol = typeChecker.getAliasedSymbol(exportedSymbol);
            }
            if (privateJsDocTag) {
                // TODO: to add a Set to keep the internal originalSymbol, if it's referenced, we put it into the NE namespace. 
                if ((exportedSymbol.flags & ts.SymbolFlags.Alias) && (originalSymbol.flags & ts.SymbolFlags.Module)) {
                    // We need to detect tag on exported symbol with alias flag.
                    const parentNode = exportedSymbol.declarations?.[0]?.parent?.parent;
                    if (parentNode) {
                        const tags = ts.getJSDocTags(parentNode).map(tag => {return { name: tag.tagName.escapedText } as ts.JSDocTagInfo});
                        if (hasJsDocTag(tags, privateJsDocTag)) {
                            continue;
                        }
                    }
                } else {
                    if (hasJsDocTag(originalSymbol.getJsDocTags(), privateJsDocTag)) {
                        continue;
                    }
                }
            }

            if ((exportedSymbol.getFlags() & ts.SymbolFlags.Prototype) &&
                (exportedSymbol.getFlags() & ts.SymbolFlags.Property)) {
                // A symbol with both class and namespace declaration
                // might have a prototype exported symbol, which associates no declarations.
                continue;
            }

            let symbolInfo = exportMap.get(originalSymbol);
            if (symbolInfo === undefined) {
                symbolInfo = {
                    exportPorts: [],
                };
                if (originalSymbol.getFlags() & ts.SymbolFlags.Module) {
                    symbolInfo.children = [];
                    const nestedModule: distributeExports.InternalModuleMeta = {
                        symbol: moduleSymbol,
                        mainExports: [],
                        aliasExports: [],
                        [prioritySymbol]: moduleMeta[prioritySymbol],
                    };
                    symbolInfo.children.push(nestedModule);
                    iterateModuleExports(originalSymbol, nestedModule);
                }

                exportMap.set(originalSymbol, symbolInfo);
            }

            symbolInfo.exportPorts.push({
                through: exportedSymbol,
                module: moduleMeta,
            });
        }
    }

    function findBestExportMeta(originalSymbol: ts.Symbol, exportPorts: ExportPort[]): number {
        // If there is only one export port, that's it.
        if (exportPorts.length === 1) {
            return 0;
        }

        // If any of the ports is specified with priority, we take the hightest specified one.
        const iHighestPriorityPort = exportPorts
            .map((_, index) => index)
            .sort((a, b) => exportPorts[a].module[prioritySymbol] - exportPorts[b].module[prioritySymbol])[0];
        if (!isNonSpecifiedPriority(exportPorts[iHighestPriorityPort].module[prioritySymbol], parsedPriorityList)) {
            return iHighestPriorityPort;
        }

        // Otherwise, We first search if there is an export is specified as 'main' by user.
        const iMatched = exportPorts.findIndex((exportPort) => matchExportPort(exportPort));
        if (iMatched >= 0) {
            return iMatched;
        }
        
        // If not, we prefer the module which exports the original.
        const iOriginal = exportPorts.findIndex((exportPort) => exportPort.through === originalSymbol);
        if (iOriginal >= 0) {
            return iOriginal;
        }

        // If no module exports original, we use the first we met.
        return 0;
    }

    function matchExportPort(exportPort: ExportPort): boolean {
        return false; // TODO
    }
}

const prioritySymbol = Symbol('Priority');

function getExportPriority(moduleSymbol: ts.Symbol, parsedPriorityList: string[]): number {
    const index = parsedPriorityList.indexOf(moduleSymbol.getName());
    return index >= 0 ? index : parsedPriorityList.length;
}

function isNonSpecifiedPriority(priority: number, parsedPriorityList: string[]) {
    return priority === parsedPriorityList.length;
}

export namespace distributeExports {
    export interface ModuleMeta extends InternalModuleMeta {
    }
    
    export interface InternalModuleMeta {
        symbol: ts.Symbol;
        mainExports: Array<{
            originalSymbol: ts.Symbol;
            exportSymbol: ts.Symbol;
            children?: InternalModuleMeta[];
        }>;
        aliasExports: Array<{
            module: InternalModuleMeta;

            /**
             * Index to the `mainExports` of `module`.
             */
            mainExportIndex: number;
            exportSymbol: ts.Symbol;
        }>;

        /**
         * Index into the priority list indicates what priority this module has, to export a symbol.
         * If this module is not in priority list, it's set to length of the priority list.
         */
        [prioritySymbol]: number;
    }

    export type ManualMainExport = string | RegExp;
}
