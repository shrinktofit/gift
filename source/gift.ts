
import * as fs from 'fs-extra';
import * as path from 'path';
import ts from 'typescript';
import ps from 'path';
import * as rConcepts from './r-concepts';
import { NameResolver } from './name-resolver';
import { distributeExports } from './distribute-exports';
import { recastTopLevelModule } from './recast';

export interface IOptions {
    input: string | string[];
    rootDir?: string;
    output: string;

    name?: string;
    rootModule?: string;
    entries?: Record<string, string>;

    exportPrivates?: string;
    shelterName?: string;
    verbose?: boolean;
}

export interface IBundleResult {
    error: GiftErrors;
    typeReferencePaths?: string[];
    code?: string;
}

export enum GiftErrors {
    Ok,
    InputFileNotFound,
    RootModuleAbsent,
    Fatal,
}

export function bundle(options: IOptions): IBundleResult {
    if (options.verbose) {
        console.log(`Cwd: ${process.cwd()}`);
        console.log(`Options: ${JSON.stringify(options)}`);
    }

    // Check the input.
    const inputs = Array.isArray(options.input) ? options.input : [options.input];
    if (!inputs.every(input => fs.existsSync(input))) {
        return { error: GiftErrors.InputFileNotFound };
    }

    return rollupTypes(options);
}

class SymbolEntityMap {
    public set(symbol: ts.Symbol, entity: rConcepts.Entity) {
        // return this._map.set(symbol, entity);
        (symbol as any)[this._entitySymbol] = entity;
    }

    public get(symbol: ts.Symbol): rConcepts.Entity | undefined {
        // return this._map.get(symbol);
        return (symbol as any)[this._entitySymbol];
    }

    // private _map: Map<ts.Symbol, rConcepts.Entity> = new Map();

    private _entitySymbol = Symbol('[[Entity]]');
}

export function rollupTypes(options: IOptions) {
    const inputs = Array.isArray(options.input) ? options.input : [options.input];
    const entries = getEntries();
    const program = createProgram();
    const typeChecker = program.getTypeChecker();
    const statements = bundle();
    return emit(statements);

    function getEntries() {
        if (options.entries) {
            return options.entries;
        } else if (options.rootModule && options.name) {
            return {
                [options.name]: options.rootModule,
            };
        }
        throw new Error(`'entries' is not specified.`);
    }

    function createTscOptions(): ts.CompilerOptions {
        return {
            rootDir: options.rootDir ?? ps.dirname(inputs[0]),
        };
    }

    function createProgram(): ts.Program {
        const tscOptions = createTscOptions();
        return ts.createProgram({
            rootNames: inputs,
            options: tscOptions,
        });
    }

    function bundle() {
        const ambientModules = typeChecker.getAmbientModules();

        const entryModules = Object.entries(entries).map(([entryModuleName, entryModuleId]) => {
            const name = `"${entryModuleId}"`;
            const moduleSymbol = ambientModules.find(m => m.getName() === name);
            if (!moduleSymbol) {
                throw new Error(`Entry ${entryModuleName}: ${entryModuleId} is not found.`);
            }
            return {
                name: entryModuleName,
                symbol: moduleSymbol,
            };
        });

        const rEntityMap = new SymbolEntityMap();

        const exportDistribution = distributeExports(entryModules.map((eM) => eM.symbol), typeChecker, []);

        const distributionMap = new Map<distributeExports.InternalModuleMeta, rConcepts.NamespaceTraits>();

        const neNamespaceMap = new Map<rConcepts.NamespaceTraits, {
            ns: rConcepts.NamespaceTraits;
            statements: ts.Statement[];
        }>();

        const rExternalModules = entryModules.map((entryModule, iEntryModule) => {
            const rModule = rConcepts.createModule(entryModule.name, entryModule.symbol);
            createREntities(exportDistribution[iEntryModule], rModule.namespaceTraits!);
            return rModule;
        });

        const visitModules = (
            moduleMeta: distributeExports.ModuleMeta,
            fx: (moduleMeta: distributeExports.ModuleMeta) => void,
        ) => {
            fx(moduleMeta);
            for (const mainExport of moduleMeta.mainExports) {
                if (mainExport.children) {
                    for (const child of mainExport.children) {
                        visitModules(child, fx);
                    }
                }
            }
        };

        for (const distribution of exportDistribution) {
            visitModules(distribution, addAliasExports);
        }

        const nameResolver = new NameResolver();
        const myRecast = (rModule: rConcepts.ModuleTraits) => recastTopLevelModule({
            program,
            typeChecker,
            rModule,
            nameResolver,
            resolveEntity: (symbol) => rEntityMap.get(symbol),
            registerNonExportedSymbol,
        });

        const statements: ts.Statement[] = [];
        for (const rModule of rExternalModules) {
            statements.push(...myRecast(rModule.moduleTraits!));
        }
        return statements;

        function createREntities(
            moduleExportDistribution: distributeExports.InternalModuleMeta,
            parent: rConcepts.NamespaceTraits,
        ) {
            distributionMap.set(moduleExportDistribution, parent);
            return moduleExportDistribution.mainExports.forEach((mainExport) => {
                const rEntity = new rConcepts.Entity(parent, mainExport.exportSymbol.name, mainExport.originalSymbol);
                if (mainExport.children) {
                    const namespaceTraits = rEntity.addNamespaceTraits();
                    for (const nestedModule of mainExport.children) {
                        createREntities(nestedModule, namespaceTraits);
                    }
                }
                rEntityMap.set(mainExport.originalSymbol, rEntity);
                return rEntity;
            });
        }

        function addAliasExports(moduleDistribution: distributeExports.InternalModuleMeta) {
            const rModule = distributionMap.get(moduleDistribution)!;
            for (const aeDistribution of moduleDistribution.aliasExports) {
                rModule.addAliasExport({
                    module: distributionMap.get(aeDistribution.module)!,
                    importName: aeDistribution.module.mainExports[aeDistribution.mainExportIndex].exportSymbol.name,
                    exportName: aeDistribution.exportSymbol.name,
                });
            }
        }

        function registerNonExportedSymbol(symbol: ts.Symbol, currentNamespace: rConcepts.NamespaceTraits) {
            let currentNamespaceInSource = currentNamespace;
            while (!currentNamespaceInSource.entity.symbol) {
                currentNamespaceInSource = currentNamespaceInSource.entity.parent.entity.namespaceTraits!;
            }

            // const neNamespaceParent = currentExportingModule;
            const neNamespaceParent = currentNamespaceInSource.entity.ownerModuleOrThis;
            let neNamespace = neNamespaceParent.namespaceTraits.neNamespace;
            if (!neNamespace) {
                neNamespace = neNamespaceParent.namespaceTraits.createNENamespace();
            }
            const name = generateUniqueName(symbol, neNamespace.trait, currentNamespaceInSource);
            const entity = new rConcepts.Entity(neNamespace.trait, name, symbol);
            rEntityMap.set(symbol, entity);
            return {
                entity,
                addStatements: (statements: ts.Statement[]) => {
                    neNamespace!.statements.push(...statements);
                },
            };
        }

        function generateUniqueName(symbol: ts.Symbol, parentModule: rConcepts.NamespaceTraits, referenceNamespaceTraits: rConcepts.NamespaceTraits): string {
            const declaration0 = symbol.getDeclarations()?.[0];
            if (!declaration0) {
                return symbol.getName();
            }

            const namespaces: string[] = [];
            let current: ts.Node = declaration0;
            while (true) {
                if (ts.isSourceFile(current)) {
                    break;
                } else if (ts.isModuleDeclaration(current) && ts.isSourceFile(current.parent)) {
                    namespaces.push(createIdFromModuleName(current.name));
                }
                current = current.parent;
            }
            
            if (!(ts.isModuleDeclaration(declaration0) && ts.isSourceFile(declaration0.parent))) {
                namespaces.push(symbol.getName());
            }
            return namespaces.join('_');
        }

        function createIdFromModuleName(name: ts.ModuleName) {
            if (ts.isIdentifier(name)) {
                return name.text;
            } else {
                return name.text.replace(/[\/\\-]/g, '_');
            }
        }
    }

    function emit(statements: ts.Statement[]) {
        const outputPath = options.output;

        // if (options.verbose) {
        //     console.log(`Referenced source files:[`);
        //     for (const referencedSourceFile of this._referencedSourceFiles) {
        //         console.log(`  ${referencedSourceFile.fileName},`);
        //     }
        //     console.log(`]`);
        // }

        const printer = ts.createPrinter({
            newLine: ts.NewLineKind.LineFeed,
        });
        const sourceFile = ts.createSourceFile(
            path.basename(outputPath),
            '',
            ts.ScriptTarget.Latest,
            false,
            ts.ScriptKind.TS,
        );

        const lines: string[] = [];
        // const typeReferencePaths: string[] = [];
        // if (rootModule.declarations && rootModule.declarations.length !== 0) {
        //     const declaration0 = rootModule.declarations[0];
        //     const rootSourceFile = declaration0.getSourceFile();
        //     const resolvedTypeReferenceDirectives: ts.Map<ts.ResolvedTypeReferenceDirective> =
        //         this._program.getResolvedTypeReferenceDirectives();
        //     resolvedTypeReferenceDirectives.forEach((trd, key) => {
        //         if (!trd.resolvedFileName) {
        //             return;
        //         }
        //         const trdSourceFile = this._program.getSourceFile(trd.resolvedFileName);
        //         if (!trdSourceFile || !this._referencedSourceFiles.has(trdSourceFile)) {
        //             return;
        //         }
        //         lines.push(`/// <reference types="${key}"/>`);
        //         typeReferencePaths.push(trd.resolvedFileName);
        //     });
        // }
        const statementsArray = ts.createNodeArray(statements);
        const result = printer.printList(
            ts.ListFormat.MultiLine, statementsArray, sourceFile);
        lines.push(result);
//         const expandTypeReferenceDirectives: Record<string, string> = {};
//         const copyTypeReferenceDirectives: string[] = [];

//         const typeReferencePaths: string[] = [];
//         if (rootModule.declarations && rootModule.declarations.length !== 0) {
//             const declaration0 = rootModule.declarations[0];
//             const indexSourceFile = declaration0.getSourceFile();
//             const indexSourceFileName = indexSourceFile.fileName;
//             for (const typeReferenceDirective of indexSourceFile.typeReferenceDirectives) {
//                 const resolveResult = ts.resolveTypeReferenceDirective(
//                     typeReferenceDirective.fileName,
//                     indexSourceFileName,
//                     this._tsCompilerOptions, {
//                         fileExists: ts.sys.fileExists,
//                         readFile: ts.sys.readFile,
//                     });
//                 if (!resolveResult.resolvedTypeReferenceDirective ||
//                     resolveResult.resolvedTypeReferenceDirective.packageId ||
//                     resolveResult.resolvedTypeReferenceDirective.primary ||
//                     resolveResult.resolvedTypeReferenceDirective.isExternalLibraryImport ||
//                     !resolveResult.resolvedTypeReferenceDirective.resolvedFileName) {
//                     copyTypeReferenceDirectives.push(typeReferenceDirective.fileName);
//                 } else {
//                     expandTypeReferenceDirectives[typeReferenceDirective.fileName] = resolveResult.resolvedTypeReferenceDirective.resolvedFileName;
//                 }
//             }
//         }

//         const lines: string[] = [];
//         for (const trd of copyTypeReferenceDirectives) {
//             lines.push(`/// <reference types="${trd}"/>`);
//         }

//         const statementsArray = ts.createNodeArray(statements);
//         const result = printer.printList(ts.ListFormat.None, statementsArray, sourceFile);
//         lines.push(result);

//         for (const trd of Object.keys(expandTypeReferenceDirectives)) {
// //             const referencedSource = fs.readFileSync(expandTypeReferenceDirectives[trd]).toString();
// //             lines.push(`
// // /// Included from type reference directive ${trd}.

// // ${referencedSource}
// // `);
//             typeReferencePaths.push(trd);
//         }
        const code = lines.join('\n');
        return { error: GiftErrors.Ok, code };
    }
}
