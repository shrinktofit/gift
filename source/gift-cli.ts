#!/usr/bin/env node

import * as fs from 'fs-extra';
import * as path from 'path';
import * as yargs from 'yargs';
import { bundle, IOptions } from './gift';

main();

async function main() {
    yargs.demandOption([ 'i', 'r' ]);
    yargs.option('input', {
        alias: 'i',
        description: 'The input files(that contains `declare module "..." { }`).',
        array: true,
    });
    yargs.option('root-dir', {
        type: 'string',
        description: 'The root dir.'
    });
    yargs.alias('r', 'root').describe('r', 'The root module name.');
    yargs.alias('n', 'name').describe('n', 'The generated module name.');
    yargs.alias('o', 'output').describe('o', 'The output file.');
    yargs.alias('u', 'shelter-name').describe(
        'u', 'Name of the unexported symbols\' namespace.(defaulted to "__internal")');
    yargs.alias('p', 'export-privates').describe(
        'p', 'Indicates whether export private members of class.');
    yargs.option('verbose', { type: 'boolean', default: false });
    yargs.option('config', {
        type: 'string',
    });
    yargs.help();

    const argv = yargs.parse(process.argv);
    const { i, n, o, r, u, p, verbose, config: configFile, rootDir } = argv;

    let entries: Record<string, string> | undefined;
    if (configFile) {
        let config;
        try {
            config = fs.readJsonSync(configFile as string);
        } catch (err) {
            console.error(`Failed to read config file ${configFile}\n`, err);
        }
        entries = config.entries;
    }

    let name: undefined | string;
    if (typeof n === 'string') {
        name = n;
    } else if (typeof o === 'string') {
        name = path.basename(o, path.extname(o));
    } else {
        console.error(`You must specify a name for the result module.`);
        return -1;
    }

    let output: undefined | string;
    if (typeof o !== 'string') {
        output = path.join('.', 'gitf-out', `${name}.d.ts`);
    } else {
        if (fs.existsSync(o) && fs.statSync(o).isDirectory() ||
            !o.toLocaleLowerCase().endsWith('.d.ts')) {
            output = path.join(o, `${name}.d.ts`);
        } else {
            output = o;
        }
    }

    const options: IOptions = {
        input: i as string,
        rootDir: rootDir as (string | undefined),
        name,
        output,
        rootModule: r as string,
        shelterName: u as (string | undefined),
        exportPrivates: p as (string | undefined),
        verbose: verbose as boolean,
        entries,
    };

    try {
        const bundleResult = bundle(options);
        await Promise.all(bundleResult.groups.map(async (group) => {
            await fs.outputFile(group.path, group.code, { encoding: 'utf8' });
        }));
    } catch (err) {
        console.error(err);
        return -1;
    }

    return 0;
}
