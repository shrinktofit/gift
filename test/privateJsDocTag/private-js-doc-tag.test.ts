
import ps from 'path';
import { bundle } from '../../source/gift';

test('Private JS doc tag', async () => {
    const inputPath = ps.join(__dirname, './test-case/index.d.ts');
    const { groups } = bundle({
        input: inputPath,
        rootModule: inputPath,
        name: 'testTag',
        output: ps.join(__dirname, './outout.d.ts'),
        privateJsDocTag: 'internal',        
    });
    expect(groups.length).toBe(1);
    expect(groups[0].code).toMatchSnapshot();
});