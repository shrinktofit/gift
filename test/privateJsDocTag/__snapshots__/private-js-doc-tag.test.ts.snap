// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Private JS doc tag 1`] = `
"declare module \\"testTag\\" {
    export namespace publicNS {
        export const publicVar: number;
    }
    export const publicNamedExport: number;
    export class A {
        publicMethod(): __private._index__InternalA;
        get publicAccessor();
        set publicAccessor(v: __private._index__InternalA);
        publicField: number;
    }
    export const typeLiternal: {
        publicMethod(): __private._index__InternalA;
        publicProp: number;
    };
    export interface publicInterface {
        publicMethod(): __private._index__internalInterface;
        publicProp: number;
    }
    export const publicFromAll: number;
    export namespace __private {
        /**
         * @internal
         */
        export class _index__InternalA {
        }
        /**
         * @internal
         */
        export interface _index__internalInterface {
            publicMethod(): _index__InternalA;
            publicProp: number;
        }
    }
    export {};
}
"
`;
